Move bookmark module to apypie

ephracis opened the pull request:
Refactor the foreman_bookmark module to use apypie instead of nailgun
so we can remove a lot of the boilerplate and keep the code clean.

See #274

ephracis commented:
I had to add support for `state: present_with_defaults` but took the liberty to also allow for `state: created` as mentioned in #146

mdellweg commented on modules/foreman_bookmark.py:
```diff
@@ -156,17 +146,19 @@ def main():
+    try:
+        search = 'name="{}",controller="{}"'.format(entity_dict['name'], entity_dict['controller'])
+        entity = module.find_resource('bookmarks', search, failsafe=True)
+    except Exception as e:
```
Heads up, we are about to be getting rid of that try block:
https://github.com/theforeman/foreman-ansible-modules/pull/281

ephracis replied:
Ok, maybe I'll wait for that one to be merged then. Guess it would be best to get #279 in as well. 🙂

mdellweg replied:
Well, that one is hard to predict...

evgeni replied:
I wouldn't wait on #279 -- that can take some time, and is "easy" as it's just a `git mv`.

#281 is merged tho :)

ephracis replied:
Cool! Removing the try-catch then.

mdellweg commented:
Seems like you are getting a hang of this...

ephracis commented:
> Seems like you are getting a hang of this...

Yup, tiny steps, one at a time. 😉

mdellweg commented on module_utils/foreman_helper.py:
```diff
@@ -209,7 +209,10 @@ def ensure_resource(self, resource, entity_dict, entity, state, check_missing=No
         changed, changed_entity = False, entity
 
-        if state == 'present':
+        if state == 'present_with_defaults' or state == 'created':
```
I guess, this will be up for discussion.
@sean797 @evgeni @Fobhep

sean797 replied:
I would suggest removing `created` for now and keeping these modules consistent. Perhaps we can open another PR and separately changing this across all the modules?

I think I'd prefer not to support both `present_with_defaults` & `created` but I'm happy with whatever if it's supported by others.

ephracis replied:
Ok, I removed it.

mdellweg commented on modules/foreman_bookmark.py:
```diff
@@ -32,8 +32,8 @@
   - "Bernhard Hopfenmueller (@Fobhep) ATIX AG"
+  - "Christoffer Reijer (@ephracis) Basalt AB"
 requirements:
-  - "nailgun >= 0.29.0"
```
:tada: every time i see it.

evgeni replied:
ACK, but probably this needs a `apypie` here now?

ephracis replied:
Done.

evgeni commented on module_utils/foreman_helper.py:
```diff
@@ -209,7 +209,10 @@ def ensure_resource(self, resource, entity_dict, entity, state, check_missing=No
         changed, changed_entity = False, entity
 
-        if state == 'present':
+        if state == 'present_with_defaults':
```
Thanks! I was too lazy to add that when initially doing the Apypie backend, because I didn't *need* it ;)

evgeni commented:
> I had to add support for `state: present_with_defaults` but took the liberty to also allow for `state: created` as mentioned in #146

I don't see the `created` in the code?

ephracis commented:
> > I had to add support for `state: present_with_defaults` but took the liberty to also allow for `state: created` as mentioned in #146
> 
> I don't see the `created` in the code?

That's correct. I added it to the backend but didn't add it all the way up to the module. But I removed it all together now after the discussion. Let it be added in another PR (if we will even go with `created` as the name, it was just a suggestion from another issue).

evgeni commented:
ok, that makes sense, just wanted to make sure I didn't miss anything :)

Fobhep commented:
thanks for the refactor @ephracis :)
